src/GitHub.Exports/Extensions/SimpleRepositoryModelExtensions.cs

```diff
@@ -30,5 +35,58 @@ public static bool MightContainSolution(this ISimpleRepositoryModel repository)
                       .Any(x => ((x.Attributes.HasFlag(FileAttributes.Directory) || x.Attributes.HasFlag(FileAttributes.Normal)) &&
                                 !x.Name.StartsWith(".", StringComparison.Ordinal) && !x.Name.StartsWith("readme", StringComparison.OrdinalIgnoreCase)));
         }
+
+        public static string CurrentSha(this ISimpleRepositoryModel repository)
+        {
+            var repo = GitService.GitServiceHelper.GetRepo(repository.LocalPath);
+            return repo.Commits.FirstOrDefault()?.Sha;
+        }
+
+        public static Uri GenerateUrl(this ISimpleRepositoryModel repository, string path = null,
+            int startLine = -1, int endLine = -1)
+        {
+            var cloneUrl = repository.CloneUrl;
+            if (cloneUrl == null)
+                return null;
+
+            if (path != null && Path.IsPathRooted(path))
+            {
+                // if the path root doesn't match the repository local path, then ignore it
+                if (repository.LocalPath == null ||
+                    !path.StartsWith(repository.LocalPath, StringComparison.OrdinalIgnoreCase))
+                {
+                    Debug.Assert(false, String.Format(CultureInfo.CurrentCulture, "GenerateUrl: path {0} doesn't match repository {1}", path, repository.LocalPath));
+                    path = null;
+                }
+                else
+                    path = path.Substring(repository.LocalPath.Length + 1);
+            }
+
+            var commitSha = repository.CurrentSha();
+            var uri = GenerateUrl(cloneUrl.ToRepositoryUrl().AbsoluteUri, commitSha, path, startLine, endLine);
+            return new UriString(uri).ToUri();
+        }
+
+        static string GenerateUrl(string basePath, string sha, string path, int startLine = -1, int endLine = -1)
+        {
+            var sb = new StringBuilder(basePath);
```

Comment:
For small number of string concatenations, I usually don't choose to use `StringBuilder` because the overhead of it isn't paid off until you get to a certain number of concatenations.

Given this is not performance critical code (it's not called often and not in a tight loop) I think it probably doesn't matter either way.

Jon Skeet has a great blog post about the subject. http://jonskeet.uk/csharp/stringbuilder.html